### PR TITLE
PEP621: Migrate setup.py config into pyproject.toml

### PR DESCRIPTION
**Pull request by cclauss**
## Proposed changes

Describe the big picture of your changes here to communicate to the maintainers
why we should accept this pull request. If it fixes a bug or resolves a feature
request, be sure to link to that issue.

[PEP621](https://peps.python.org/pep-0621/) encourages the Python community to migrate away from complex `setup.cfg` + `setup.py` files by putting config data in `pyproject.toml`.

Migrate `setup.py` to `setup.cfg` using [setuptools-py2cfg](https://pypi.org/project/setuptools-py2cfg) plus manual modifications.
Then migrate `setup.cfg` to `pyproject.toml` using [ini2toml](https://pypi.org/project/ini2toml) to do the file conversion and running [validate-pyproject](https://github.com/abravalheri/validate-pyproject) in pre-commit to validate the results.
* https://setuptools.pypa.io/en/latest/userguide/pyproject_config.html
* https://packaging.python.org/en/latest/tutorials/packaging-projects/#creating-pyproject-toml

The change in `tox.ini` from `pip install --no-binary lxml` --> `pip install --no-cache-dir lxml` is going to be mandated by the next major release of `pip` and decreases our tox test runtimes from 3 or 4 minutes down to 1 or 2 minutes.

Building Python projects in this way requires dropping support for Python 3.6.
* https://devguide.python.org/versions

If this pull request is merged then [__required__](https://docs.github.com/en/actions/using-workflows/required-workflows) must be removed from `test (3.6)` and added to `test (3.11)`.
Also, maybe remove required from `flake8` and add it to `pre-commit.ci`?

Screenshot 2023-02-26 at 23 56 25

## Types of changes

Please check the type of change your PR introduces:

- [ ] Release (new release request)
- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Code style update (PEP8, lint, formatting, renaming, etc)
- [x] Refactoring (no functional changes, no api changes)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [x] Build-related changes (build process, tests runner, etc)
- [ ] Other (please describe):

## Checklist

_Put an `x` in the boxes that apply. You can also fill these out after creating
the PR. If you're unsure about any of them, don't hesitate to ask. We're here to
help! This is simply a reminder of what we are going to look for before merging
your code._

- [ ] Lint and unit tests pass locally with my changes
- [x] I have added tests that prove my fix is effective or that my feature works

## Other information

Any other information that is important to this PR such as screenshots of how
the component looks before and after the change.

### CODE REVIEW

**Comment by nijel** [comment]
> The change in tox.ini from pip install --no-binary lxml --> pip install --no-cache-dir lxml is going to be mandated by the next major release of pip and decreases our tox test runtimes from 3 or 4 minutes down to 1 or 2 minutes.

We really need `--no-binary` there, see https://github.com/SAML-Toolkits/python3-saml/issues/292, https://github.com/python-social-auth/social-core/issues/659, https://github.com/python-social-auth/social-core/pull/738

File: pyproject.toml [hunk 1]
```diff
@@ -0,0 +1,99 @@
...
+]
+testing = [
+  "coverage>=3.6",
+  "cryptography>=2.1.1",
+  "httpretty>=0.9.6",
+  "pytest>=4.5",
+  "pytest-cov>=2.7.1",
+  "python-jose>=3",
+  "python3-saml>=1.5"
+]
```
Comment (nijel) [inline pyproject.toml:73]: This all duplicates requirements-*.txt content. Can the dependencies be specified just once?

Comment (cclauss) [reply pyproject.toml:73]: Yes, but can we push that to a separate PR?  This one already touches a lot of files and `tox.ini`, and GitHub Action caching will need to be modified when dropping requirements files.

Comment (davidbgk) [reply pyproject.toml:73]: You can generate the requirements files from the `pyproject.toml` one:

```
python3 -m piptools compile --output-file=requirements.txt pyproject.toml
python3 -m piptools compile --extra=dev --output-file=requirements-dev.txt pyproject.toml
etc
```

**Pull request closed without merging**